pymc3/distributions/multivariate.py

```diff
@@ -1790,119 +1805,83 @@ class KroneckerNormal(Continuous):
     ----------
     .. [1] Saatchi, Y. (2011). "Scalable inference for structured Gaussian process models"
     """
+    rv_op = kroneckernormal
 
-    def __init__(self, mu, covs=None, chols=None, evds=None, sigma=None, *args, **kwargs):
-        self._setup(covs, chols, evds, sigma)
-        super().__init__(*args, **kwargs)
-        self.mu = at.as_tensor_variable(mu)
-        self.mean = self.median = self.mode = self.mu
+    @classmethod
+    def dist(cls, mu, covs=None, chols=None, evds=None, sigma=None, *args, **kwargs):
 
-    def _setup(self, covs, chols, evds, sigma):
-        self.cholesky = Cholesky(lower=True, on_error="raise")
+        cholesky = Cholesky(lower=True, on_error="raise")
         if len([i for i in [covs, chols, evds] if i is not None]) != 1:
             raise ValueError(
                 "Incompatible parameterization. Specify exactly one of covs, chols, or evds."
             )
-        self._isEVD = False
-        self.sigma = sigma
-        self.is_noisy = self.sigma is not None and self.sigma != 0
+
+        eigh_iterable = []
+        is_noisy = sigma is not None and sigma != 0
+
         if covs is not None:
-            self._cov_type = "cov"
-            self.covs = covs
-            if self.is_noisy:
+            cov = kronecker(*covs)
+            if is_noisy:
                 # Noise requires eigendecomposition
                 eigh_map = map(eigh, covs)
-                self._setup_evd(eigh_map)
+                eigh_iterable = eigh_map
+                cov = cov + sigma ** 2 * at.identity_like(cov)
             else:
                 # Otherwise use cholesky as usual
-                self.chols = list(map(self.cholesky, self.covs))
-                self.chol_diags = list(map(at.diag, self.chols))
-                self.sizes = at.as_tensor_variable([chol.shape[0] for chol in self.chols])
-                self.N = at.prod(self.sizes)
+                chols = list(map(cholesky, covs))
+                chols = [_chol.eval() for _chol in chols]
```

Comment:
I don't know how practical it is to use an `eval()` in a `dist`. But this is one way to get the output's of the cholesky matrices from the `Op`.

Reply:
Yeah, we almost never want to use `eval`/`aesara.function` outside of a place where it's intended to be used (e.g. posterior predictive sampling).

Regardless, why would we want/need to use non-symbolic values here?

Reply:
The exact place this comment tags right now i.e. `dist` is `eval()` free. The `.eval()` is currently required over in `rng_fn` since the `kronecker()` method was returning `TensorVariable` (I'm working on replacing that with `scipy.linalg.kron` so we can count on `eval()` being removed from there too) The other place `.eval()` is in is `logp` where it is used to infer shapes (Theoretically that can also be removed with a bit of effort.)

Overall with the current implementation (which is a whole lot of omitted code) I wouldn't worry about `eval()` as much as the inability of supporting `sigma` which us the bigger issue with this implementation.

Reply:
We simply cannot evaluate things like this unless we place very strong and unnatural restrictions on _all_ of the terms involved.  More specifically, the graphs in `chols` must have no symbolic inputs; otherwise, this will fail.  If the tests are passing with this in place, then our testing isn't sufficient, because it's only ever testing `Constant`/`SharedVariable` inputs.